Sources/PactVerificationService.swift

```diff
@@ -76,108 +78,199 @@ open class PactVerificationService {
 
-  fileprivate func setupInteractions (_ interactions: [Interaction]) -> Future<String, NSError> {
-    let promise = Promise<String, NSError>()
-    let payload: [String: Any] = ["interactions": interactions.map({ $0.payload() }),
-                                  "example_description": "description"]
+  func handle(result: Result<String, NSError>, completion: @escaping VoidHandler) {
+    switch result {
+    case .success:
+      completion(.success(()))
+    case .failure(let error):
+      completion(.failure(error))
+    }
+  }
 
-    self.performNetworkRequest(for: Router.setup(payload), promise: promise)
+  func handle(result: Result<String, URLSession.APIServiceError>, completion: @escaping VoidHandler) {
+    switch result {
+    case .success:
+      completion(.success(()))
+    case .failure(let error):
+      completion(.failure(NSError.prepareWith(message: error.localizedDescription)))
+    }
+  }
 
-    return promise.future
+  func handle(result: Result<String, URLSession.APIServiceError>, completion: @escaping StringHandler) {
+    switch result {
+    case .success(let resultString):
+      completion(.success(resultString))
+    case .failure(let error):
+      completion(.failure(NSError.prepareWith(message: error.localizedDescription)))
+    }
   }
 
-  // MARK: - Networking
+}
+
+// MARK: - Network request handler
 
-  private let session = URLSession(configuration: URLSessionConfiguration.ephemeral)
+fileprivate extension PactVerificationService {
 
-  private func performNetworkRequest(for router: Router, promise: Promise<String, NSError>) {
-    let task: URLSessionDataTask?
+  var session: URLSession {
+    URLSession(configuration: URLSessionConfiguration.ephemeral)
+  }
+
+  func performNetworkRequest(for router: Router, completion: @escaping (Result<String, URLSession.APIServiceError>) -> Void) {
     do {
-      task = try session.dataTask(with: router.asURLRequest()) { data, response, error in
-        self.responseHandler(promise)(data, response, error)
+      let dataTask = try session.dataTask(with: router.asURLRequest()) { result in
+        switch result {
+        case .success(let (response, data)):
+          guard let statusCode = (response as? HTTPURLResponse)?.statusCode, 200..<299 ~= statusCode else {
+            completion(.failure(.invalidResponse(NSError.prepareWith(data: data))))
+            return
+          }
+          guard let responseString = String(data: data, encoding: .utf8) else {
+            completion(.failure(.noData))
+            return
+          }
+          completion(.success(responseString))
+        case .failure(let error):
+          completion(.failure(.apiError(error)))
+        }
       }
-
-      task?.resume()
+      dataTask.resume()
     } catch {
-      DispatchQueue.main.async {
-        // Make sure this promise fails in the future.
-        promise.failure(error as NSError)
-      }
+      completion(.failure(.invalidEndpoint))
```

Comment:
I don't like this as it could be any exception triggering an invalidEndpoint type of error. I understand why it is being done though and maybe we can clean it up in a future PR (i.e. remove the throwing of the Error at line 49).

Reply:
Nice catch (pun intended)